solidity/contracts/Peggy.sol

```diff
@@ -343,6 +347,152 @@ contract Peggy {
 		}
 	}
 
+	// This is being used purely to avoid stack too deep errors
+	struct LogicCallArgs {
+		// Transfers out to the logic contract
+		uint256[] transferAmounts;
+		address[] transferTokenContracts;
+		// The fees (transferred to msg.sender)
+		uint256[] feeAmounts;
+		address[] feeTokenContracts;
+		// The arbitrary logic call
+		address logicContractAddress;
+		bytes payload;
+		// Invalidation metadata
+		uint256 timeOut;
+		bytes32 invalidationId;
+		uint256 invalidationNonce;
+	}
+
+	// This makes calls to contracts that execute arbitrary logic
+	// First, it gives the logic contract some tokens
+	// Then, it gives msg.senders tokens for fees
+	// Then, it calls an arbitrary function on the logic contract
+	// invalidationId and invalidationNonce are used for replay prevention.
+	// They can be used to implement a per-token nonce by setting the token
+	// address as the invalidationId and incrementing the nonce each call.
+	// They can be used for nonce-free replay prevention by using a different invalidationId
+	// for each call.
+	function submitLogicCall(
+		// The validators that approve the call
+		address[] memory _currentValidators,
+		uint256[] memory _currentPowers,
+		uint256 _currentValsetNonce,
+		// These are arrays of the parts of the validators signatures
+		uint8[] memory _v,
+		bytes32[] memory _r,
+		bytes32[] memory _s,
+		LogicCallArgs memory _args
+	) public {
+		// CHECKS scoped to reduce stack depth
+		{
+			// Check that the call has not timed out
+			require(block.timestamp < _args.timeOut, "Timed out");
```

Comment:
can we be consistent and use block numbers as established in #170 there are no real fundamental advantages to using timestamps.

Reply:
Oh yea sorry i knew i was forgetting something